<img align="left" src="https://ithaka-labs.s3.amazonaws.com/static-files/images/tdm/tdmdocs/CC_BY.png"><br />

This notebook is adapted by Zhuo Chen from the notebooks created by [Nathan Kelber](https://github.com/ithaka/tdm-notebooks/blob/e6275296c010280909e90e3ea47922d52d99c5a7/pandas-2.ipynb), [William Mattingly](https://github.com/wjbmattingly/tap-2022-pandas) and [Melanie Walsh](https://github.com/melaniewalsh/Data-Analysis-with-Pandas) under [Creative Commons CC BY License](https://creativecommons.org/licenses/by/4.0/)<br />
For questions/comments/improvements, email zhuo.chen@ithaka.org or nathan.kelber@ithaka.org.<br />
___

# Pandas 2

**Description:** This notebook describes how to:

* Index a dataframe
* Sort a dataframe
* Work with missing values in a dataframe
* Filter data in a dataframe
* Update a dataframe

This is the second notebook in a series on learning to use Pandas. 

**Use Case:** For Learners (Detailed explanation, not ideal for researchers)

**Difficulty:** Intermediate

**Knowledge Required:** 
* [Pandas 1](./pandas-1.ipynb)
* Python Basics ([Start Python Basics I](./python-basics-1.ipynb))

**Knowledge Recommended:** 
* [Python Intermediate 2](./python-intermediate-2.ipynb)
* [Python Intermediate 4](./python-intermediate-4.ipynb)

**Completion Time:** 90 minutes

**Data Format:** CSV (.csv)

**Libraries Used:** Pandas

**Research Pipeline:** None
___


In [1]:
# Import pandas library, `as pd` allows us to shorten typing `pandas` to `pd` when we call pandas
import pandas as pd

## Indexing a dataframe

In this section, we will continue working with the dataframe we created in Pandas 1 storing data on the most recent 10 World Cup games. 

In [2]:
# Create a dataframe with world cup data
wcup = pd.DataFrame({"Year": [2022, 
                              2018, 
                              2014, 
                              2010, 
                              2006, 
                              2002, 
                              1998, 
                              1994, 
                              1990,
                              1986], 
                     "Champion": ["Argentina", 
                                  "France", 
                                  "Germany", 
                                  "Spain", 
                                  "Italy", 
                                  "Brazil", 
                                  "France", 
                                  "Brazil", 
                                  "Germany", 
                                  "Argentina"], 
                     "Host": ["Qatar", 
                              "Russia", 
                              "Brazil", 
                              "South Africa", 
                              "Germany", 
                              "Korea/Japan", 
                              "France", 
                              "USA", 
                              "Italy", 
                              "Mexico"],
                     "Score": ["7-5", 
                               "4-2", 
                               "1-0", 
                               "1-0", 
                               "6-4", 
                               "2-0", 
                               "3-0", 
                               "3-2", 
                               "1-0", 
                               "3-2"]
                    })
wcup['Goals Scored'] = wcup['Score'].apply(lambda r: r.split('-')[0])
wcup['Goals Conceded'] = wcup['Score'].apply(lambda r: r.split('-')[1])
wcup['Difference'] = wcup['Goals Scored'].astype(int) - wcup['Goals Conceded'].astype(int)
wcup

Year   Champion          Host Score Goals Scored Goals Conceded  Difference
0  2022  Argentina         Qatar   7-5            7              5           2
1  2018     France        Russia   4-2            4              2           2
2  2014    Germany        Brazil   1-0            1              0           1
3  2010      Spain  South Africa   1-0            1              0           1
4  2006      Italy       Germany   6-4            6              4           2
5  2002     Brazil   Korea/Japan   2-0            2              0           2
6  1998     France        France   3-0            3              0           3
7  1994     Brazil           USA   3-2            3              2           1
8  1990    Germany         Italy   1-0            1              0           1
9  1986  Argentina        Mexico   3-2            3              2           1

### Set, reset and use indexes
We have seen that by default, the rows in a dataframe are numbered by integer indexes starting from 0. The indexes look like a column to the far left without a name. 

We can set the index column to one of the columns in the dataframe. This is desirable because a range of integers is not descriptive but a column with a name is descriptive. When we want to locate specific data, descriptive labels are much more useful. 

In [3]:
# Set index column to 'Year'
wcup.set_index('Year')

Champion          Host Score Goals Scored Goals Conceded  Difference
Year                                                                       
2022  Argentina         Qatar   7-5            7              5           2
2018     France        Russia   4-2            4              2           2
2014    Germany        Brazil   1-0            1              0           1
2010      Spain  South Africa   1-0            1              0           1
2006      Italy       Germany   6-4            6              4           2
2002     Brazil   Korea/Japan   2-0            2              0           2
1998     France        France   3-0            3              0           3
1994     Brazil           USA   3-2            3              2           1
1990    Germany         Italy   1-0            1              0           1
1986  Argentina        Mexico   3-2            3              2           1

Take a look at the original dataframe, is it changed? 

In [4]:
# Take a look at the original dataframe
wcup

Year   Champion          Host Score Goals Scored Goals Conceded  Difference
0  2022  Argentina         Qatar   7-5            7              5           2
1  2018     France        Russia   4-2            4              2           2
2  2014    Germany        Brazil   1-0            1              0           1
3  2010      Spain  South Africa   1-0            1              0           1
4  2006      Italy       Germany   6-4            6              4           2
5  2002     Brazil   Korea/Japan   2-0            2              0           2
6  1998     France        France   3-0            3              0           3
7  1994     Brazil           USA   3-2            3              2           1
8  1990    Germany         Italy   1-0            1              0           1
9  1986  Argentina        Mexico   3-2            3              2           1

The original dataframe is **NOT** changed after we use the `.set_index()` method to change the index column. This is because in Pandas, we have a distinction between a view and a copy. When a view of the dataframe is returned, any change we make will affect the original dataframe, but when a copy is returned, any change we make only affects the copy, not the original dataframe. The `.set_index()` method returns a copy, this is why the original dataframe is not affected.  

If you want to make the change permanent, you can assign the returned object to the variable where you store the original dataframe to update it. 

In [5]:
# Change the index column and commit the change
wcup = wcup.set_index('Year')
wcup

Champion          Host Score Goals Scored Goals Conceded  Difference
Year                                                                       
2022  Argentina         Qatar   7-5            7              5           2
2018     France        Russia   4-2            4              2           2
2014    Germany        Brazil   1-0            1              0           1
2010      Spain  South Africa   1-0            1              0           1
2006      Italy       Germany   6-4            6              4           2
2002     Brazil   Korea/Japan   2-0            2              0           2
1998     France        France   3-0            3              0           3
1994     Brazil           USA   3-2            3              2           1
1990    Germany         Italy   1-0            1              0           1
1986  Argentina        Mexico   3-2            3              2           1

## Sort a dataframe

We can also sort the dataframe by the index column. Here, we have a numerical column as our index colummn. When we sort the indexes, by default, the dataframe will be sorted by the index column in an ascending order. 

In [6]:
# Sort the indexes
wcup.sort_index()

Champion          Host Score Goals Scored Goals Conceded  Difference
Year                                                                       
1986  Argentina        Mexico   3-2            3              2           1
1990    Germany         Italy   1-0            1              0           1
1994     Brazil           USA   3-2            3              2           1
1998     France        France   3-0            3              0           3
2002     Brazil   Korea/Japan   2-0            2              0           2
2006      Italy       Germany   6-4            6              4           2
2010      Spain  South Africa   1-0            1              0           1
2014    Germany        Brazil   1-0            1              0           1
2018     France        Russia   4-2            4              2           2
2022  Argentina         Qatar   7-5            7              5           2

You could set the parameter `ascending=False` to sort the indexes in a descending order.

In [7]:
# Specify the ascending order
wcup.sort_index(ascending=False)

Champion          Host Score Goals Scored Goals Conceded  Difference
Year                                                                       
2022  Argentina         Qatar   7-5            7              5           2
2018     France        Russia   4-2            4              2           2
2014    Germany        Brazil   1-0            1              0           1
2010      Spain  South Africa   1-0            1              0           1
2006      Italy       Germany   6-4            6              4           2
2002     Brazil   Korea/Japan   2-0            2              0           2
1998     France        France   3-0            3              0           3
1994     Brazil           USA   3-2            3              2           1
1990    Germany         Italy   1-0            1              0           1
1986  Argentina        Mexico   3-2            3              2           1

### Sort by one column

We can sort the entire dataframe by a column other than the index column. The `.sort_values()` method helps us do it. 

In [8]:
# Sort the dataframe by the column 'Goals Scored'
wcup.sort_values(by=['Goals Scored'])

Champion          Host Score Goals Scored Goals Conceded  Difference
Year                                                                       
2014    Germany        Brazil   1-0            1              0           1
2010      Spain  South Africa   1-0            1              0           1
1990    Germany         Italy   1-0            1              0           1
2002     Brazil   Korea/Japan   2-0            2              0           2
1998     France        France   3-0            3              0           3
1994     Brazil           USA   3-2            3              2           1
1986  Argentina        Mexico   3-2            3              2           1
2018     France        Russia   4-2            4              2           2
2006      Italy       Germany   6-4            6              4           2
2022  Argentina         Qatar   7-5            7              5           2

### Sort by multiple columns
It is a convention to sort the soccer results first by difference (i.e. how many more goals the champion scored than the runner-up) and then by goals conceded (i.e. how many goals the champion lost). Pandas can easily do that. 

In [9]:
# Sort the dataframe by Difference column in descending order 
# then by Goals Conceded column in ascending order
wcup.sort_values(by=['Difference', 'Goals Conceded'], ascending=[False, True])

Champion          Host Score Goals Scored Goals Conceded  Difference
Year                                                                       
1998     France        France   3-0            3              0           3
2002     Brazil   Korea/Japan   2-0            2              0           2
2018     France        Russia   4-2            4              2           2
2006      Italy       Germany   6-4            6              4           2
2022  Argentina         Qatar   7-5            7              5           2
2014    Germany        Brazil   1-0            1              0           1
2010      Spain  South Africa   1-0            1              0           1
1990    Germany         Italy   1-0            1              0           1
1994     Brazil           USA   3-2            3              2           1
1986  Argentina        Mexico   3-2            3              2           1

In [10]:
# Reset the index for later use
wcup = wcup.reset_index()

### Hierarchical indexing
Sorting by multiple columns helps us group the data in a certain way. For example, in the world cup dataframe, if we would like to group the data first by the champions and then by the years, we can sort the dataframe by these two columns and set the two columns as the a multi-level index of the dataframe. The `Champion` column will be the level 0 index and the `Year` column will be the level 1 index. The hierarchical indexing allows us to work with higher dimension data.

In [11]:
# First,sort the two columns we will use for multi-level indexing
# Then, set the index to the composite of Champion and Year
wcup = wcup.sort_values(by=['Champion', 'Year'], ascending=[1,0]).set_index(['Champion', 'Year'])

In [12]:
# Take a look at the updated dataframe
wcup

Host Score Goals Scored Goals Conceded  Difference
Champion  Year                                                            
Argentina 2022         Qatar   7-5            7              5           2
          1986        Mexico   3-2            3              2           1
Brazil    2002   Korea/Japan   2-0            2              0           2
          1994           USA   3-2            3              2           1
France    2018        Russia   4-2            4              2           2
          1998        France   3-0            3              0           3
Germany   2014        Brazil   1-0            1              0           1
          1990         Italy   1-0            1              0           1
Italy     2006       Germany   6-4            6              4           2
Spain     2010  South Africa   1-0            1              0           1

If a dataframe has a multi-level index, to access a certain row, we will need to provide a multi-level index in order to access it. 

In [13]:
# Get the data on the 1986 world cup game won by Argentina
wcup.loc[('Argentina', 1986)]

Host              Mexico
Score                3-2
Goals Scored           3
Goals Conceded         2
Difference             1
Name: (Argentina, 1986), dtype: object

## A quick review of how to create a dataframe from a file

In [Pandas 1](./pandas-1.ipynb), we learned how to create a dataframe by reading in a csv or an excel file. 

For example, we can convert the data in a .csv file to a Pandas DataFrame using the `.read_csv()` method. We pass in the location of the .csv file.

In [16]:
### Download the sample file for this Lesson
import urllib
url = 'https://ithaka-labs.s3.amazonaws.com/static-files/images/tdm/tdmdocs/Pandas1_failed_banks_since_2000.csv'
path = './data/' + url.rsplit('/', 1)[-1]
urllib.request.urlretrieve(url, path)
print('Sample file retrieved.')

Sample file retrieved.


Use the `**File > Open**` menu above to navigate to the `failed_banks_since_2000.csv` in the `/data` folder. Preview its structure before we load it into a dataframe.

In [17]:
# Create a DataFrame `df` from a CSV file using the .read_csv() method
df = pd.read_csv(path) # pass in the location of the file
df

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
..                                 ...                ...   ...    ...   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539  
2                            Equity Bank    23-Oct-20  10538  
3              United Fidelity Bank, fsb    16-Oct-20  10537  
4                         MVB Bank, Inc.     3-Apr-20  10536  
..                                   ...          ...    ...  
560                Superior Federal, FSB    27-Jul-01   6004  
561                    North Valley Bank     3-May-01   4648  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
563              Banterra Bank of Marion    14-Dec-00   4646  
564                   Bank of the Orient    13-Oct-00   4645  

[565 rows x 7 columns]

By default, Pandas displays the first five rows and the last five rows of the dataframe. You can change the display setting using the `.set_option()` method.

In [18]:
# Change the display setting
pd.set_option('display.min_rows', 20) # set the minimum number of rows to display to 20
df

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
555                       NextBank, NA            Phoenix    AZ  22314   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539  
2                            Equity Bank    23-Oct-20  10538  
3              United Fidelity Bank, fsb    16-Oct-20  10537  
4                         MVB Bank, Inc.     3-Apr-20  10536  
5             Farmers and Merchants Bank    14-Feb-20  10535  
6                        Industrial Bank     1-Nov-19  10534  
7                     Buckeye State Bank    25-Oct-19  10533  
8      Kentucky Farmers Bank Corporation    25-Oct-19  10532  
9                     Legend Bank, N. A.    31-May-19  10531  
..                                   ...          ...    ...  
555                                  NaN     7-Feb-02   4653  
556       The State Bank & Trust Company     1-Feb-02   4652  
557     The Security State Bank of Pecos    18-Jan-02   4651  
558     Israel Discount Bank of New York    11-Jan-02   4650  
559                   Delta Trust & Bank     7-Sep-01   4649  
560                Superior Federal, FSB    27-Jul-01   6004  
561                    North Valley Bank     3-May-01   4648  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
563              Banterra Bank of Marion    14-Dec-00   4646  
564                   Bank of the Orient    13-Oct-00   4645  

[565 rows x 7 columns]

The display setting is global throughout the notebook. Therefore, any dataframe in the current notebook will have this setting in effect.

Now, you see that Pandas displays the first 10 rows and the last 10 rows of the dataframe.

By convention, a dataframe variable is called `df` but we could give it any valid Python variable name. Here, we follow the convention. 

In [19]:
# Get some info about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 565 entries, 0 to 564
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Bank Name              565 non-null    object
 1   City                   565 non-null    object
 2   State                  565 non-null    object
 3   Cert                   565 non-null    int64 
 4   Acquiring Institution  534 non-null    object
 5   Closing Date           565 non-null    object
 6   Fund                   565 non-null    int64 
dtypes: int64(2), object(5)
memory usage: 31.0+ KB


The `info()` method tells us that there are 565 rows and 7 columns in the dataframe. Almost all columns have 565 non-null values, except the column of `Acquiring Institution`.

In the next code cell, can you index the dataframe `df` using a multi-level index where the level 0 index is the `State` and level 1 index is the `City`?

In [ ]:
# set a hierarchical index for df


<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

In the exercises in this notebook, we'll work on a dataset built from Constellate.

We'll use the `constellate` client to automatically retrieve the [metadata](https://constellate.org/docs/key-terms/#metadata) for a [dataset](https://constellate.org/docs/key-terms/#dataset). We can retrieve the [metadata](https://constellate.org/docs/key-terms/#metadata) in a [CSV file](https://constellate.org/docs/key-terms/#csv-file) using the `get_metadata` method.


In [20]:
# Creating a variable `dataset_id` to hold our dataset ID
# The default dataset is Shakespeare Quarterly, 1950-present
# retrieve the metadata
import constellate
dataset_id = "7e41317e-740f-e86a-4729-20dab492e925"
metadata = constellate.get_metadata(dataset_id)

Constellate: use and download of datasets is covered by the Terms & Conditions of Use: https://constellate.org/terms-and-conditions/
All documents from JSTOR published in Shakespeare Quarterly from 1950 - 2020. 6745 documents.
INFO:root:File /Users/zchen/data/7e41317e-740f-e86a-4729-20dab492e925-sampled-metadata.csv exists. Not re-downloading.


In [21]:
# Check the type of the value stored in metadata
type(metadata)

str

In [22]:
# print out the string stored in metadata
print(metadata)

/Users/zchen/data/7e41317e-740f-e86a-4729-20dab492e925-sampled-metadata.csv


The metadata is stored in a .csv file. In the following code cell, read in the data using Pandas. Give the dataframe a name other than `df`. Then print out the dataframe to take a look. 

In [23]:
# Convert the csv file to a dataframe


id  \
0      http://www.jstor.org/stable/2902131   
1      http://www.jstor.org/stable/2871217   
2      http://www.jstor.org/stable/2866912   
3     http://www.jstor.org/stable/41819770   
4      http://www.jstor.org/stable/2868793   
5      http://www.jstor.org/stable/2869261   
6      http://www.jstor.org/stable/2870776   
7      http://www.jstor.org/stable/2869933   
8      http://www.jstor.org/stable/2866221   
9      http://www.jstor.org/stable/2870918   
...                                    ...   
1490  http://www.jstor.org/stable/41679774   
1491   http://www.jstor.org/stable/2868044   
1492   http://www.jstor.org/stable/2870754   
1493   http://www.jstor.org/stable/2868513   
1494   http://www.jstor.org/stable/2871404   
1495   http://www.jstor.org/stable/2869482   
1496   http://www.jstor.org/stable/2867034   
1497   http://www.jstor.org/stable/2871079   
1498   http://www.jstor.org/stable/2868190   
1499   http://www.jstor.org/stable/2866411   

                                                  title  \
0      Fragments of Nationalism in Troilus and Cressida   
1     Shakespearean Chronology, Ideological Complici...   
2                                        Review Article   
3                                        Review Article   
4                                        Review Article   
5        Falstaff in Windsor Forest: Villain or Victim?   
6            Coriolanus: Body Politic and Private Parts   
7                     Shakespeare in West Germany, 1983   
8                                        Review Article   
9                                        Review Article   
...                                                 ...   
1490                                     Review Article   
1491                                     Review Article   
1492                      They Were Not Such Good Years   
1493                      The Third Murderer in Macbeth   
1494                                     Review Article   
1495                           The New Arden Coriolanus   
1496                                     Review Article   
1497                                                [H]   
1498                                     Review Article   
1499                                     Review Article   

                   isPartOf  publicationYear  doi  docType provider  \
0     Shakespeare Quarterly             2000  NaN  article    jstor   
1     Shakespeare Quarterly             1994  NaN  article    jstor   
2     Shakespeare Quarterly             1962  NaN  article    jstor   
3     Shakespeare Quarterly             2012  NaN  article    jstor   
4     Shakespeare Quarterly             1974  NaN  article    jstor   
5     Shakespeare Quarterly             1975  NaN  article    jstor   
6     Shakespeare Quarterly             1990  NaN  article    jstor   
7     Shakespeare Quarterly             1984  NaN  article    jstor   
8     Shakespeare Quarterly             1951  NaN  article    jstor   
9     Shakespeare Quarterly             1992  NaN  article    jstor   
...                     ...              ...  ...      ...      ...   
1490  Shakespeare Quarterly             2012  NaN  article    jstor   
1491  Shakespeare Quarterly             1962  NaN  article    jstor   
1492  Shakespeare Quarterly             1989  NaN  article    jstor   
1493  Shakespeare Quarterly             1972  NaN  article    jstor   
1494  Shakespeare Quarterly             1997  NaN  article    jstor   
1495  Shakespeare Quarterly             1979  NaN  article    jstor   
1496  Shakespeare Quarterly             1959  NaN  article    jstor   
1497  Shakespeare Quarterly             1964  NaN  article    jstor   
1498  Shakespeare Quarterly             1963  NaN  article    jstor   
1499  Shakespeare Quarterly             1955  NaN  article    jstor   

     datePublished  issueNumber  volumeNumber  \
0       2000-07-01          2.0            51   
1       1994-07-01          2.0            45   
2       1962-01-01

Use a Pandas method to explore the dataframe. How many rows does it have? How many columns does it have? What is the data type of the data in each column?

In [ ]:
# Use a pandas method to explore the dataframe


## Filter dataframe

A common pipeline in data processing in Pandas is that you create a dataframe from a file and then reduce the dataframe only to the rows and columns that you are interested in. 

We have learned how to use `.loc` and `.iloc` to select part of a dataframe in [Pandas 1](./pandas-1.ipynb). We will learn more ways to do data filtering in this section.

### Work with missing values
It is a common case that datasets have missing values. As you may have already noticed, blank cells in a CSV file show up as NaN in a Pandas DataFrame. For example, in the dataset of failed banks, the `Acquiring Institution` column gives the name when a failed bank was acquired by another institution and is empty otherwise.

In Pandas, we have a bunch of methods that can create a boolean mask over the data. One of them is `.isna()`. This method helps us check whether a pandas series or a dataframe has missing values. 

In [24]:
# Use isna() to check whether a dataframe has missing values
df.isna()

Bank Name   City  State   Cert  Acquiring Institution  Closing Date  \
0        False  False  False  False                  False         False   
1        False  False  False  False                  False         False   
2        False  False  False  False                  False         False   
3        False  False  False  False                  False         False   
4        False  False  False  False                  False         False   
5        False  False  False  False                  False         False   
6        False  False  False  False                  False         False   
7        False  False  False  False                  False         False   
8        False  False  False  False                  False         False   
9        False  False  False  False                  False         False   
..         ...    ...    ...    ...                    ...           ...   
555      False  False  False  False                   True         False   
556      False  False  False  False                  False         False   
557      False  False  False  False                  False         False   
558      False  False  False  False                  False         False   
559      False  False  False  False                  False         False   
560      False  False  False  False                  False         False   
561      False  False  False  False                  False         False   
562      False  False  False  False                  False         False   
563      False  False  False  False                  False         False   
564      False  False  False  False                  False         False   

      Fund  
0    False  
1    False  
2    False  
3    False  
4    False  
5    False  
6    False  
7    False  
8    False  
9    False  
..     ...  
555  False  
556  False  
557  False  
558  False  
559  False  
560  False  
561  False  
562  False  
563  False  
564  False  

[565 rows x 7 columns]

The `.isna()` method puts a mask on the original dataframe. The cells with a non-null value are masked with the boolean value of `False`. The cells with a null value are masked with the boolean value of `True`.

We can also use `.isna()` to check whether a specific column has missing values. 

In [25]:
# Use isna() to check whether a column has missing values
df['Acquiring Institution'].isna()

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
       ...  
555     True
556    False
557    False
558    False
559    False
560    False
561    False
562    False
563    False
564    False
Name: Acquiring Institution, Length: 565, dtype: bool

### Drop rows and columns with missing values

If you want to exclude the rows and columns with missing values from your data analysis, you can use the `.dropna()` method to do that.

By default, the `.dropna()` method drops the rows with at least one missing value. 

In [26]:
# Use .dropna() to remove all rows with at least one missing value
df.dropna() # no argument passed in

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
554              Net 1st National Bank         Boca Raton    FL  26652   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539  
2                            Equity Bank    23-Oct-20  10538  
3              United Fidelity Bank, fsb    16-Oct-20  10537  
4                         MVB Bank, Inc.     3-Apr-20  10536  
5             Farmers and Merchants Bank    14-Feb-20  10535  
6                        Industrial Bank     1-Nov-19  10534  
7                     Buckeye State Bank    25-Oct-19  10533  
8      Kentucky Farmers Bank Corporation    25-Oct-19  10532  
9                     Legend Bank, N. A.    31-May-19  10531  
..                                   ...          ...    ...  
554                       Bank Leumi USA     1-Mar-02   4654  
556       The State Bank & Trust Company     1-Feb-02   4652  
557     The Security State Bank of Pecos    18-Jan-02   4651  
558     Israel Discount Bank of New York    11-Jan-02   4650  
559                   Delta Trust & Bank     7-Sep-01   4649  
560                Superior Federal, FSB    27-Jul-01   6004  
561                    North Valley Bank     3-May-01   4648  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
563              Banterra Bank of Marion    14-Dec-00   4646  
564                   Bank of the Orient    13-Oct-00   4645  

[534 rows x 7 columns]

You can also set the axis parameter to 0 to drop the rows with missing values.

In [27]:
# Use .dropna() to remove all rows with at least one missing value
df.dropna(axis=0) # Set the axis to 0

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
554              Net 1st National Bank         Boca Raton    FL  26652   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539  
2                            Equity Bank    23-Oct-20  10538  
3              United Fidelity Bank, fsb    16-Oct-20  10537  
4                         MVB Bank, Inc.     3-Apr-20  10536  
5             Farmers and Merchants Bank    14-Feb-20  10535  
6                        Industrial Bank     1-Nov-19  10534  
7                     Buckeye State Bank    25-Oct-19  10533  
8      Kentucky Farmers Bank Corporation    25-Oct-19  10532  
9                     Legend Bank, N. A.    31-May-19  10531  
..                                   ...          ...    ...  
554                       Bank Leumi USA     1-Mar-02   4654  
556       The State Bank & Trust Company     1-Feb-02   4652  
557     The Security State Bank of Pecos    18-Jan-02   4651  
558     Israel Discount Bank of New York    11-Jan-02   4650  
559                   Delta Trust & Bank     7-Sep-01   4649  
560                Superior Federal, FSB    27-Jul-01   6004  
561                    North Valley Bank     3-May-01   4648  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
563              Banterra Bank of Marion    14-Dec-00   4646  
564                   Bank of the Orient    13-Oct-00   4645  

[534 rows x 7 columns]

Or, you can set the axis parameter to 'rows' drop the rows with missing values.

In [28]:
# Use .dropna() to remove all rows with at least one missing value
df.dropna(axis='rows') # Set the axis to 'rows'

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
554              Net 1st National Bank         Boca Raton    FL  26652   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539  
2                            Equity Bank    23-Oct-20  10538  
3              United Fidelity Bank, fsb    16-Oct-20  10537  
4                         MVB Bank, Inc.     3-Apr-20  10536  
5             Farmers and Merchants Bank    14-Feb-20  10535  
6                        Industrial Bank     1-Nov-19  10534  
7                     Buckeye State Bank    25-Oct-19  10533  
8      Kentucky Farmers Bank Corporation    25-Oct-19  10532  
9                     Legend Bank, N. A.    31-May-19  10531  
..                                   ...          ...    ...  
554                       Bank Leumi USA     1-Mar-02   4654  
556       The State Bank & Trust Company     1-Feb-02   4652  
557     The Security State Bank of Pecos    18-Jan-02   4651  
558     Israel Discount Bank of New York    11-Jan-02   4650  
559                   Delta Trust & Bank     7-Sep-01   4649  
560                Superior Federal, FSB    27-Jul-01   6004  
561                    North Valley Bank     3-May-01   4648  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
563              Banterra Bank of Marion    14-Dec-00   4646  
564                   Bank of the Orient    13-Oct-00   4645  

[534 rows x 7 columns]

If you set the axis parameter to 1, you will drop the columns with missing values. 

In [29]:
# Use .dropna() to remove all columns with at least one missing value
df.dropna(axis=1) # set the axis to 1

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
555                       NextBank, NA            Phoenix    AZ  22314   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

    Closing Date   Fund  
0      12-Mar-23  10540  
1      10-Mar-23  10539  
2      23-Oct-20  10538  
3      16-Oct-20  10537  
4       3-Apr-20  10536  
5      14-Feb-20  10535  
6       1-Nov-19  10534  
7      25-Oct-19  10533  
8      25-Oct-19  10532  
9      31-May-19  10531  
..           ...    ...  
555     7-Feb-02   4653  
556     1-Feb-02   4652  
557    18-Jan-02   4651  
558    11-Jan-02   4650  
559     7-Sep-01   4649  
560    27-Jul-01   6004  
561     3-May-01   4648  
562     2-Feb-01   4647  
563    14-Dec-00   4646  
564    13-Oct-00   4645  

[565 rows x 6 columns]

You can also drop the columns with missing values by setting the axis parameter to 'columns'.

In [30]:
# Use .dropna() to remove all columns with at least one missing value
df.dropna(axis='columns') # set the axis to 'columns'

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
555                       NextBank, NA            Phoenix    AZ  22314   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

    Closing Date   Fund  
0      12-Mar-23  10540  
1      10-Mar-23  10539  
2      23-Oct-20  10538  
3      16-Oct-20  10537  
4       3-Apr-20  10536  
5      14-Feb-20  10535  
6       1-Nov-19  10534  
7      25-Oct-19  10533  
8      25-Oct-19  10532  
9      31-May-19  10531  
..           ...    ...  
555     7-Feb-02   4653  
556     1-Feb-02   4652  
557    18-Jan-02   4651  
558    11-Jan-02   4650  
559     7-Sep-01   4649  
560    27-Jul-01   6004  
561     3-May-01   4648  
562     2-Feb-01   4647  
563    14-Dec-00   4646  
564    13-Oct-00   4645  

[565 rows x 6 columns]

Sometimes we would want to drop a row only if that row has a missing value in a specific column. We can use the subset parameter to specify the column(s) to look for missing values. 

In [31]:
# Specify the columns to look for missing values
df.dropna(subset=['Acquiring Institution', 'City'])

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
554              Net 1st National Bank         Boca Raton    FL  26652   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539  
2                            Equity Bank    23-Oct-20  10538  
3              United Fidelity Bank, fsb    16-Oct-20  10537  
4                         MVB Bank, Inc.     3-Apr-20  10536  
5             Farmers and Merchants Bank    14-Feb-20  10535  
6                        Industrial Bank     1-Nov-19  10534  
7                     Buckeye State Bank    25-Oct-19  10533  
8      Kentucky Farmers Bank Corporation    25-Oct-19  10532  
9                     Legend Bank, N. A.    31-May-19  10531  
..                                   ...          ...    ...  
554                       Bank Leumi USA     1-Mar-02   4654  
556       The State Bank & Trust Company     1-Feb-02   4652  
557     The Security State Bank of Pecos    18-Jan-02   4651  
558     Israel Discount Bank of New York    11-Jan-02   4650  
559                   Delta Trust & Bank     7-Sep-01   4649  
560                Superior Federal, FSB    27-Jul-01   6004  
561                    North Valley Bank     3-May-01   4648  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647  
563              Banterra Bank of Marion    14-Dec-00   4646  
564                   Bank of the Orient    13-Oct-00   4645  

[534 rows x 7 columns]

Note that the `.dropna()` method only returns a copy, not a view. This means that any change you make using the `.dropna()` method will not affect the original dataframe. To make the change permanent, you can assign the result to the variable where you store the original dataframe to update it.

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

Sometimes you may want to exclude a row/column from your consideration when you decide whether to drop a row/column. In other words, even if a row/column has a missing value, you don't want to drop it. How do you do that? 

In the dataset with data on failed banks, let's say we want to drop any row with missing values except the rows in `Acquiring Institution`. In other words, we want to preserve the rows in `Acquiring Institution` no matter whether it has missing values or not. 

In [ ]:
# Drop any row with missing values except the rows in 'Acquiring Institution'


Oftentimes, you would want to maintain the rows and columns that have missing values. However, you would want to fill the cells with NaN values with some values which are of the same data type as the other cells in the same column. In this way, when you apply a certain function to a column in a dataframe, you will not run into type error. A common practice to deal with this kind of problem is to use the `.fillna()` method. 

In [ ]:
# Fill the missing values
df['Acquiring Institution'].fillna('No Acquirer')

### Drop certain columns or rows

We have seen how to drop rows or columns with missing values. Sometimes, even if a row or a column does not have a missing value, you still want to drop them because you will not use them in your analysis anyways. In this case, we will use the `.drop()` method to remove those rows or columns.

You can specify which column you want to drop using the 'columns' parameter. 

In [ ]:
# Drop a column by setting the columns parameter
df.drop(columns='Fund')

You can drop multiple columns at one time. 

In [ ]:
# Drop multiple columns by setting the columns parameter
df.drop(columns=['Fund', 'Cert'])

Another way to drop a column is to give the label of the column you want to drop and then set the axis parameter to 1.

In [ ]:
# Drop a column by setting the axis parameter
df.drop('Fund', axis=1)

You can also drop multiple columns by setting the axis parameter. 

In [ ]:
# Drop multiple columns by setting the axis parameter
df.drop(['Fund', 'Cert'], axis=1)

To drop a row, you can specify which row you want to drop using the 'index' parameter.

In [ ]:
# Drop a row by setting the index parameter
df.drop(index=0)

In the next code cell, can you write some code to drop multiple rows from df?

In [ ]:
# Drop multiple rows using the index parameter


Another way to drop a row is to give the label of the row you want to drop and then set the axis parameter to 0.

In [ ]:
# Drop a row by setting the axis parameter
df.drop(0, axis=0)

We know that by default, the rows are indexed with integer numbers. You could set the index column to one of the columns of the dataframe. In the next code cell, can you set the index column to the `State` column and then drop all rows with the label 'GA' or 'KS'?

In [ ]:
# Drop multiple rows by setting the axis parameter


You might want to drop multiple consecutive rows at one time. The `.drop()` method does not have a parameter for slicing but we can come up with a workaround.

We can use the `.index` property to get the range of indexes for the rows we want to drop and pass them to the `.drop()` method.

In [ ]:
# Drop multiple consecutive rows
df.drop(df.index[2:5], axis=0) 

To drop multiple consecutive columns, we can use the `.columns` attribute to get the range of the indexes for the columns we want to drop and then pass it to the `.drop()` method. 

In [ ]:
# Drop multiple consecutive columns
df.drop(df.columns[3:5], axis=1)

The `.drop()` method returns a copy, not a view. Therefore, whatever change you make using it will not affect the original dataframe. 

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

When you explore the Shakespeare dataframe, what did you find about the column `doi`? What did you find about the column `placeOfPublication`? Is there any non-null value in them?

In [ ]:
# Drop the columns of doi and the placeofPublication, make the change permanent


### Filter data using conditionals
Conditional selection using `df.loc[]` is a very common method to filter a dataframe. 

You write a filtering condition to filter a target column. The condition then checks, for each cell in the target column, whether it fulfills the condition or not. The results will be returned as a Series of True/False values. The `.loc` indexer then uses this Series to select the rows that have True values. 

Suppose you are interested in the banks that failed since 2000 in the state of Georgia. From the original dataframe, you would like to get all the rows of the failed banks in Georgia. How do you do it?

In [43]:
# Write a filtering condition
df['State'] == 'GA' # Create a boolean mask over the column 'State'

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
       ...  
555    False
556    False
557    False
558    False
559    False
560    False
561    False
562    False
563    False
564    False
Name: State, Length: 565, dtype: bool

In [44]:
# Assign the filtering condition to a variable
filt = (df['State'] == 'GA') # Use parenthesis for better reading

In [45]:
# Put the Series returned by the filtering condition within the hard brackets of df.loc[]
df.loc[filt]

Bank Name            City State   Cert  \
19        The Woodbury Banking Company        Woodbury    GA  11297   
24                 The Bank of Georgia  Peachtree City    GA  35259   
28   Capitol City Bank & Trust Company         Atlanta    GA  33938   
36            Eastside Commercial Bank         Conyers    GA  58125   
61                        Sunrise Bank        Valdosta    GA  58185   
63                 Douglas County Bank    Douglasville    GA  21649   
69                       Frontier Bank        LaGrange    GA  16431   
74             Hometown Community Bank       Braselton    GA  57928   
85              Jasper Banking Company          Jasper    GA  16240   
88           First Cherokee State Bank       Woodstock    GA  32711   
..                                 ...             ...   ...    ...   
495                     FirstCity Bank     Stockbridge    GA  18243   
496            Freedom Bank of Georgia        Commerce    GA  57558   
506       FirstBank Financial Services       McDonough    GA  57017   
514                   Haven Trust Bank          Duluth    GA  35379   
515       First Georgia Community Bank         Jackson    GA  34301   
518                     Community Bank      Loganville    GA  16490   
522                 Alpha Bank & Trust      Alpharetta    GA  58241   
528                     Integrity Bank      Alpharetta    GA  35469   
539                            NetBank      Alpharetta    GA  32575   
550         AmTrade International Bank         Atlanta    GA  33784   

                      Acquiring Institution Closing Date   Fund  
19                              United Bank    19-Aug-16  10521  
24                            Fidelity Bank     2-Oct-15  10516  
28      First-Citizens Bank & Trust Company    13-Feb-15  10512  
36                Community & Southern Bank    18-Jul-14  10504  
61                             Synovus Bank    10-May-13  10481  
63                      Hamilton State Bank    26-Apr-13  10476  
69                HeritageBank of the South     8-Mar-13  10471  
74         CertusBank, National Association    16-Nov-12  10466  
85                        Stearns Bank N.A.    27-Jul-12  10455  
88                Community & Southern Bank    20-Jul-12  10450  
..                                      ...          ...    ...  
495                                     NaN    20-Mar-09  10047  
496                  Northeast Georgia Bank     6-Mar-09  10044  
506                            Regions Bank     6-Feb-09  10036  
514  Branch Banking & Trust Company, (BB&T)    12-Dec-08  10027  
515                             United Bank     5-Dec-08  10025  
518                           Bank of Essex    21-Nov-08  10022  
522                      Stearns Bank, N.A.    24-Oct-08  10018  
528                            Regions Bank    29-Aug-08  10012  
539                              ING DIRECT    28-Sep-07  10001  
550                                     NaN    30-Sep-02   4657  

[93 rows x 7 columns]

Out of the rows that fulfill the filtering condition, we can further specify which columns to be returned.

In [46]:
# Specify a single column to be returned
df.loc[filt, 'Bank Name']

19          The Woodbury Banking Company
24                   The Bank of Georgia
28     Capitol City Bank & Trust Company
36              Eastside Commercial Bank
61                          Sunrise Bank
63                   Douglas County Bank
69                         Frontier Bank
74               Hometown Community Bank
85                Jasper Banking Company
88             First Cherokee State Bank
                     ...                
495                       FirstCity Bank
496              Freedom Bank of Georgia
506         FirstBank Financial Services
514                     Haven Trust Bank
515         First Georgia Community Bank
518                       Community Bank
522                   Alpha Bank & Trust
528                       Integrity Bank
539                              NetBank
550           AmTrade International Bank
Name: Bank Name, Length: 93, dtype: object

Of course, we can select muliple columns to be returned out of the filtered rows. 

In [47]:
# Specify multiple columns to be returned
df.loc[filt, ['Bank Name', 'Fund']]

Bank Name   Fund
19        The Woodbury Banking Company  10521
24                 The Bank of Georgia  10516
28   Capitol City Bank & Trust Company  10512
36            Eastside Commercial Bank  10504
61                        Sunrise Bank  10481
63                 Douglas County Bank  10476
69                       Frontier Bank  10471
74             Hometown Community Bank  10466
85              Jasper Banking Company  10455
88           First Cherokee State Bank  10450
..                                 ...    ...
495                     FirstCity Bank  10047
496            Freedom Bank of Georgia  10044
506       FirstBank Financial Services  10036
514                   Haven Trust Bank  10027
515       First Georgia Community Bank  10025
518                     Community Bank  10022
522                 Alpha Bank & Trust  10018
528                     Integrity Bank  10012
539                            NetBank  10001
550         AmTrade International Bank   4657

[93 rows x 2 columns]

Now suppose you want to get all the failed banks whose name contains the word 'community'.

In [48]:
# Get all the banks with the word 'community' in their name
filt = (df['Bank Name'].str.contains('Community'))
df.loc[filt, ['Bank Name']]

Bank Name
8                                Louisa Community Bank
17                              Harvest Community Bank
29                             Highland Community Bank
49          Texas Community Bank, National Association
52                                The Community's Bank
54                                Community South Bank
56   First Community Bank of Southwest Florida (als...
62                               Pisgah Community Bank
65                              Chipola Community Bank
72                             Westside Community Bank
..                                                 ...
441                               Community First Bank
442         Community National Bank of Sarasota County
446                             Peoples Community Bank
471                        Neighborhood Community Bank
472                     Community Bank of West Georgia
475                            Southern Community Bank
482                            Citizens Community Bank
498                            Heritage Community Bank
515                       First Georgia Community Bank
518                                     Community Bank

[63 rows x 1 columns]

#### Conjunction of multiple filtering conditions: `&`

Oftentimes, you would want to filter a dataframe based on more complex conditions. For example, suppose you would like to get the banks in GA that were closed between 2008 and 2010. How do you use `df.loc[ ]` to achieve it?

The location of the failed banks is stored in the `State` column. The closing year of the banks is stored in the `Closing Date` column. 

In [49]:
# Create the first filtering condition restricting the state
filt1 = (df['State'] == 'GA')

How to get the closing year of the banks? Recall what we have learned in [Pandas 1](./pandas-1.ipynb) about creating a new column based on an old one. How do you extract the closing year out of the column `Closing Date`? We can use the `.apply()` method.

In [50]:
# Create a new column storing the closing year of the banks
df['Closing Year'] = df['Closing Date'].apply(lambda r: r.split('-')[2])
df['Closing Year'] = df['Closing Year'].astype(int)

In [51]:
# Take a look at the dataframe
df

Bank Name               City State   Cert  \
0                       Signature Bank           New York    NY  57053   
1                  Silicon Valley Bank        Santa Clara    CA  24735   
2                    Almena State Bank             Almena    KS  15426   
3           First City Bank of Florida  Fort Walton Beach    FL  16748   
4                 The First State Bank      Barboursville    WV  14361   
5                   Ericson State Bank            Ericson    NE  18265   
6     City National Bank of New Jersey             Newark    NJ  21111   
7                        Resolute Bank             Maumee    OH  58317   
8                Louisa Community Bank             Louisa    KY  58112   
9                 The Enloe State Bank             Cooper    TX  10716   
..                                 ...                ...   ...    ...   
555                       NextBank, NA            Phoenix    AZ  22314   
556           Oakwood Deposit Bank Co.            Oakwood    OH   8966   
557              Bank of Sierra Blanca      Sierra Blanca    TX  22002   
558                  Hamilton Bank, NA              Miami    FL  24382   
559             Sinclair National Bank           Gravette    AR  34248   
560                 Superior Bank, FSB           Hinsdale    IL  32646   
561                Malta National Bank              Malta    OH   6629   
562    First Alliance Bank & Trust Co.         Manchester    NH  34264   
563  National State Bank of Metropolis         Metropolis    IL   3815   
564                   Bank of Honolulu           Honolulu    HI  21029   

                   Acquiring Institution Closing Date   Fund  Closing Year  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540            23  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539            23  
2                            Equity Bank    23-Oct-20  10538            20  
3              United Fidelity Bank, fsb    16-Oct-20  10537            20  
4                         MVB Bank, Inc.     3-Apr-20  10536            20  
5             Farmers and Merchants Bank    14-Feb-20  10535            20  
6                        Industrial Bank     1-Nov-19  10534            19  
7                     Buckeye State Bank    25-Oct-19  10533            19  
8      Kentucky Farmers Bank Corporation    25-Oct-19  10532            19  
9                     Legend Bank, N. A.    31-May-19  10531            19  
..                                   ...          ...    ...           ...  
555                                  NaN     7-Feb-02   4653             2  
556       The State Bank & Trust Company     1-Feb-02   4652             2  
557     The Security State Bank of Pecos    18-Jan-02   4651             2  
558     Israel Discount Bank of New York    11-Jan-02   4650             2  
559                   Delta Trust & Bank     7-Sep-01   4649             1  
560                Superior Federal, FSB    27-Jul-01   6004             1  
561                    North Valley Bank     3-May-01   4648             1  
562  Southern New Hampshire Bank & Trust     2-Feb-01   4647             1  
563              Banterra Bank of Marion    14-Dec-00   4646             0  
564                   Bank of the Orient    13-Oct-00   4645             0  

[565 rows x 8 columns]

In [52]:
# Create the second filtering condition restricting the closing year
filt2 = (df['Closing Year'] > 7) & (df['Closing Year'] < 11)

With the two filtering conditions, we are ready to extract the banks in GA that failed between 2008 and 2010.

In [53]:
# Use filt1 and filt2 to get the target rows
df.loc[filt1 & filt2]

Bank Name           City State   Cert  \
218              United Americas Bank, N.A.        Atlanta    GA  35065   
219         Appalachian Community Bank, FSB    McCaysville    GA  58495   
220                    Chestatee State Bank    Dawsonville    GA  34578   
228                  Darby Bank & Trust Co.        Vidalia    GA  14580   
229                  Tifton Banking Company         Tifton    GA  57831   
237  The First National Bank of Barnesville    Barnesville    GA   2119   
238                         The Gordon Bank         Gordon    GA  33904   
250                        The Peoples Bank         Winder    GA    182   
251           First Commerce Community Bank   Douglasville    GA  57448   
252                         Bank of Ellijay        Ellijay    GA  58197   
269                  Northwest Bank & Trust        Acworth    GA  57658   
275         Crescent Bank and Trust Company         Jasper    GA  27559   
288                     First National Bank       Savannah    GA  34152   
304                  Satilla Community Bank    Saint Marys    GA  35114   
333                     Unity National Bank   Cartersville    GA  34678   
335                McIntosh Commercial Bank     Carrollton    GA  57399   
338                       Bank of Hiawassee      Hiawassee    GA  10054   
339              Appalachian Community Bank        Ellijay    GA  33989   
341                   Century Security Bank         Duluth    GA  58104   
360                Community Bank and Trust       Cornelia    GA   5702   
363          First National Bank of Georgia     Carrollton    GA  16480   
379              RockBridge Commercial Bank        Atlanta    GA  58315   
386                       The Tattnall Bank     Reidsville    GA  12080   
387            First Security National Bank       Norcross    GA  26290   
388             The Buckhead Community Bank        Atlanta    GA  34663   
397                    United Security Bank         Sparta    GA  22286   
412                    American United Bank  Lawrenceville    GA  57794   
418                           Georgian Bank        Atlanta    GA  57151   
434                       First Coweta Bank         Newnan    GA  57702   
435                                   ebank        Atlanta    GA  34682   
449           Security Bank of Jones County           Gray    GA   8486   
450         Security Bank of Houston County          Perry    GA  27048   
451            Security Bank of Bibb County          Macon    GA  27367   
452            Security Bank of North Metro      Woodstock    GA  57105   
453           Security Bank of North Fulton     Alpharetta    GA  57430   
454        Security Bank of Gwinnett County        Suwanee    GA  57346   
459                     First Piedmont Bank         Winder    GA  34594   
471             Neighborhood Community Bank         Newnan    GA  35285   
472          Community Bank of West Georgia     Villa Rica    GA  57436   
475                 Southern Community Bank   Fayetteville    GA  35251   
483                      Silverton Bank, NA        Atlanta    GA  26535   
487                  American Southern Bank       Kennesaw    GA  57943   
492                      Omni National Bank        Atlanta    GA  22238   
495                          FirstCity Bank    Stockbridge    GA  18243   
496                 Freedom Bank of Georgia       Commerce    GA  57558   
506            FirstBank Financial Services      McDonough    GA  57017   
514                        Haven Trust Bank         Duluth    GA  35379   
515            First Georgia Community Bank        Jackson    GA  34301   
518                          Community Bank     Loganville    GA  16490   
522                      Alpha Bank & Trust     Alpharetta    GA  58241   
528                          Integrity Bank     Alpharetta    GA  35469   

                           Acquiring Institution Closing Date   Fund  \
218                 State Bank and Trust Company    17-Dec-10

Note that when we extract rows that fulfill multiple conditions, we use `&` in Pandas, not `and`. If you replace `&` with `and`, you will get an error. This is different than what we have learned about boolean operators in [Python basics 2](./python-basics-2.ipynb). In Python, we use `and`, `or` and `not`. In Pandas, we use `&`, `|` and `~` intead. 

|Pandas Operator|Boolean|Requires|
|---|---|---|
|&|and|All required to `True`|
|\||or|If any are `True`|
|~|not|The opposite|

Although we use different symbols for these boolean operators, the truth table for them stays the same. For a quick review of the truth table, see [Python basics 2](./python-basics-2.ipynb).

#### Disjunction of multiple filtering conditions: `|`
Suppose you would like to take a look at all the failed banks in the state of Georgia or the state of New York. How do you use `df.loc[ ]` to get the target rows?

In [54]:
# Create the two filtering conditions restricting the state to GA and NY
filt1 = (df['State'] == 'GA')
filt2 = (df['State'] == 'NY')

In [55]:
# Use filt1 and filt2 to get the target rows
df.loc[filt1|filt2]

Bank Name            City State   Cert  \
0                       Signature Bank        New York    NY  57053   
19        The Woodbury Banking Company        Woodbury    GA  11297   
24                 The Bank of Georgia  Peachtree City    GA  35259   
28   Capitol City Bank & Trust Company         Atlanta    GA  33938   
36            Eastside Commercial Bank         Conyers    GA  58125   
61                        Sunrise Bank        Valdosta    GA  58185   
63                 Douglas County Bank    Douglasville    GA  21649   
69                       Frontier Bank        LaGrange    GA  16431   
74             Hometown Community Bank       Braselton    GA  57928   
85              Jasper Banking Company          Jasper    GA  16240   
..                                 ...             ...   ...    ...   
496            Freedom Bank of Georgia        Commerce    GA  57558   
506       FirstBank Financial Services       McDonough    GA  57017   
514                   Haven Trust Bank          Duluth    GA  35379   
515       First Georgia Community Bank         Jackson    GA  34301   
518                     Community Bank      Loganville    GA  16490   
522                 Alpha Bank & Trust      Alpharetta    GA  58241   
528                     Integrity Bank      Alpharetta    GA  35469   
539                            NetBank      Alpharetta    GA  32575   
542                      Reliance Bank    White Plains    NY  26778   
550         AmTrade International Bank         Atlanta    GA  33784   

                      Acquiring Institution Closing Date   Fund  Closing Year  
0               Signature Bridge Bank, N.A.    12-Mar-23  10540            23  
19                              United Bank    19-Aug-16  10521            16  
24                            Fidelity Bank     2-Oct-15  10516            15  
28      First-Citizens Bank & Trust Company    13-Feb-15  10512            15  
36                Community & Southern Bank    18-Jul-14  10504            14  
61                             Synovus Bank    10-May-13  10481            13  
63                      Hamilton State Bank    26-Apr-13  10476            13  
69                HeritageBank of the South     8-Mar-13  10471            13  
74         CertusBank, National Association    16-Nov-12  10466            12  
85                        Stearns Bank N.A.    27-Jul-12  10455            12  
..                                      ...          ...    ...           ...  
496                  Northeast Georgia Bank     6-Mar-09  10044             9  
506                            Regions Bank     6-Feb-09  10036             9  
514  Branch Banking & Trust Company, (BB&T)    12-Dec-08  10027             8  
515                             United Bank     5-Dec-08  10025             8  
518                           Bank of Essex    21-Nov-08  10022             8  
522                      Stearns Bank, N.A.    24-Oct-08  10018             8  
528                            Regions Bank    29-Aug-08  10012             8  
539                              ING DIRECT    28-Sep-07  10001             7  
542                        Union State Bank    19-Mar-04   4664             4  
550                                     NaN    30-Sep-02   4657             2  

[99 rows x 8 columns]

If you would like to get the data of the failed banks in the following six states --- Georgia, New York, New Jersey, Florida, California and West Virginia, you will not want to write six filtering conditions and use the vertical bar `|` to connect all of them. That would be too repetitive. In this case, we can use the `.isin()` method to create a filtering condition.

In [56]:
# Create a list of the states
states = ['GA', 'NY', 'NJ', 'FL', 'CA', 'WV']

In [57]:
# Create a filtering condition
filt = (df['State'].isin(states))

In [58]:
# Use filt to find all failed banks in the six states
df.loc[filt]

Bank Name               City State   Cert  \
0                           Signature Bank           New York    NY  57053   
1                      Silicon Valley Bank        Santa Clara    CA  24735   
3               First City Bank of Florida  Fort Walton Beach    FL  16748   
4                     The First State Bank      Barboursville    WV  14361   
6         City National Bank of New Jersey             Newark    NJ  21111   
17                  Harvest Community Bank         Pennsville    NJ  34951   
19            The Woodbury Banking Company           Woodbury    GA  11297   
24                     The Bank of Georgia     Peachtree City    GA  35259   
28       Capitol City Bank & Trust Company            Atlanta    GA  33938   
30        First National Bank of Crestview          Crestview    FL  17557   
..                                     ...                ...   ...    ...   
531                First Heritage Bank, NA      Newport Beach    CA  57961   
533                           IndyMac Bank           Pasadena    CA  29730   
539                                NetBank         Alpharetta    GA  32575   
542                          Reliance Bank       White Plains    NY  26778   
543  Guaranty National Bank of Tallahassee        Tallahassee    FL  26838   
544                    Dollar Savings Bank             Newark    NJ  31330   
547                  Southern Pacific Bank           Torrance    CA  27094   
550             AmTrade International Bank            Atlanta    GA  33784   
554                  Net 1st National Bank         Boca Raton    FL  26652   
558                      Hamilton Bank, NA              Miami    FL  24382   

                   Acquiring Institution Closing Date   Fund  Closing Year  
0            Signature Bridge Bank, N.A.    12-Mar-23  10540            23  
1       Silicon Valley Bridge Bank, N.A.    10-Mar-23  10539            23  
3              United Fidelity Bank, fsb    16-Oct-20  10537            20  
4                         MVB Bank, Inc.     3-Apr-20  10536            20  
6                        Industrial Bank     1-Nov-19  10534            19  
17   First-Citizens Bank & Trust Company    13-Jan-17  10523            17  
19                           United Bank    19-Aug-16  10521            16  
24                         Fidelity Bank     2-Oct-15  10516            15  
28   First-Citizens Bank & Trust Company    13-Feb-15  10512            15  
30                        First NBC Bank    16-Jan-15  10510            15  
..                                   ...          ...    ...           ...  
531                 Mutual of Omaha Bank    25-Jul-08  10009             8  
533                    OneWest Bank, FSB    11-Jul-08  10007             8  
539                           ING DIRECT    28-Sep-07  10001             7  
542                     Union State Bank    19-Mar-04   4664             4  
543              Hancock Bank of Florida    12-Mar-04   4663             4  
544                                  NaN    14-Feb-04   6006             4  
547                            Beal Bank     7-Feb-03   4660             3  
550                                  NaN    30-Sep-02   4657             2  
554                       Bank Leumi USA     1-Mar-02   4654             2  
558     Israel Discount Bank of New York    11-Jan-02   4650             2  

[227 rows x 8 columns]

#### Negation of a certain condition:`~`
Now, suppose you would like to get all the failed banks that were **not** closed in 2008. How do you do it?

In [59]:
# Create the filtering condition restricting the closing year to non-2008
filt = (~(df['Closing Year'] == 8))

In [60]:
# Use the filtering condition to get the target rows with specified columns
df.loc[filt, ['Bank Name', 'City']]

Bank Name               City
0                       Signature Bank           New York
1                  Silicon Valley Bank        Santa Clara
2                    Almena State Bank             Almena
3           First City Bank of Florida  Fort Walton Beach
4                 The First State Bank      Barboursville
5                   Ericson State Bank            Ericson
6     City National Bank of New Jersey             Newark
7                        Resolute Bank             Maumee
8                Louisa Community Bank             Louisa
9                 The Enloe State Bank             Cooper
..                                 ...                ...
555                       NextBank, NA            Phoenix
556           Oakwood Deposit Bank Co.            Oakwood
557              Bank of Sierra Blanca      Sierra Blanca
558                  Hamilton Bank, NA              Miami
559             Sinclair National Bank           Gravette
560                 Superior Bank, FSB           Hinsdale
561                Malta National Bank              Malta
562    First Alliance Bank & Trust Co.         Manchester
563  National State Bank of Metropolis         Metropolis
564                   Bank of Honolulu           Honolulu

[540 rows x 2 columns]

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

Let's do some filtering!

From the Shakespeare dataframe, get the title and the creator of the documents published between 2000 **and** 2010.

From the Shakespeare dataframe, get the creator of the documents shorter than 10 pages **or** longer than 50 pages. 

From the Shakespeare dataframe, get the title of the documents whose publisher is **not** Folger Shakespeare Library. 

## Update a dataframe
We can make changes to the data in a dataframe.
### Update headers
We can update the column names of a dataframe.

In [ ]:
# Take a look at the columns
df.columns

In [ ]:
# Access a column using the dot notation
df.City

In [ ]:
# If a column name has a space in it
df.Bank Name

We could replace all the spaces in column names with an `_`. In this way, we can access all the columns using the dot notation.

In [ ]:
# Replace spaces in column names with underscores
df.columns = df.columns.str.replace(' ', '_')
df

You could also change the case of the headers.

In [ ]:
# Change all headers to upper case
df.columns.str.upper()

We have been updating the column names all at one time. However, oftentimes we just want to update specific columns. In this case, we could use the `df.rename()` method and pass in a **dictionary** where the keys are the original column names and the values are the new column names.

In [ ]:
# Change the column name of 'CERT' to 'CERTIFICATE_NUM'
df.rename(columns = {'Cert':'Certificate_Num'})

To change multiple column names, we just pass in a dictionary to `df.rename()` with multiple key:value pairs.

In [ ]:
# Change multiple column names
df.rename(columns = {'Cert':'Certificate_Num', 'Fund':'Financial_Institution_Num'})

In [ ]:
# Make the change permanent
df = df.rename(columns = {'Cert':'Certificate_Num', 'Fund':'Financial_Institution_Num'})

### Update rows 
How to update the values in a row? In [Pandas 1](./pandas-1.ipynb), we have learned how to look up values using `.loc` and `.iloc`.

To update a row, we could use `.loc` or `.iloc` to locate it and then assign the new values to that row.

In [ ]:
# Change an entire row
df.loc[0] = ['Almena State Bank', 'Almena', 'KS', 15426, 'Equity Bank', '23-Oct-20', 10000, 20]

You can locate a specific cell in a row and update the value in that cell alone.

In [ ]:
# Change a specific value in a row
df.loc[0, 'Financial_Institution_Num'] = 10001
df

We could change multiple specific values in a row using `.loc[]`. 

In [ ]:
# Change multiple values in a row
df.loc[0, ['Bank_Name', 'Financial_Institution_Num']] = ['Almena Bank', 12000]
df

### Update columns
There are multiple methods we can use to update columns. Let's take a look at two methods `.map()` and `replace()`.

In [ ]:
# Use .map() to update specific values in a column
df['Bank_Name'].map({'Almena Bank': 'Almena State Bank', 'The First State Bank': 'West Virginia Bank'})

In [ ]:
# Use .replace() to update specific values in a column while maintaining the rest
df['Bank_Name'].replace({'Almena Bank': 'Almena State Bank', 'The First State Bank': 'West Virginia Bank'})

We can also use a filtering condition to locate the target columns and then make changes. 

For example, we can locate all the banks that failed in 2020 and change their closing date to 'Recent'.

In [ ]:
# Make a filtering condition to get the banks that failed in 2020
filt = (df['Closing_Year'] == 2020)

In [ ]:
# Use the filtering condition to locate the columns and update them
df.loc[filt, ['Financial_Institution_Num', 'Closing_Year']] = [1000, 'Recent']
df

<h3 style="color:red; display:inline">Coding Challenge! &lt; / &gt; </h3>

Make all column names in the Shakespeare dataframe upper case. 

Get all documents whose current title is 'Review Article' and change their title to 'Review'.

Get all documents whose word count exceeds 5000 and change their word count to the string 'Long article'.

___
## Lesson Complete

Congratulations! You have completed *Pandas 2*.

### Start Next Lesson: [Pandas 3 ->](./pandas-3.ipynb)

### Exercise Solutions
Here are a few solutions for exercises in this lesson.

In [ ]:
# Read in the metadata
shake = pd.read_csv(metadata)

In [ ]:
# Set the rows to display to 30
pd.set_option('display.max_rows', 30)

In [ ]:
# Explore the dataframe
shake.info()

In [ ]:
### Drop the rows and columns with at least 2 missing values

# Get the tuple (# of rows, # of columns) 
df.shape

# Store the num of rows and num of columns in two variables
num_rows = df.shape[0]
num_cols = df.shape[1]

# Drop all columns which have at least 2 missing values
df.dropna(thresh=num_rows-1, axis=1)

In [ ]:
# Drop the columns of doi and the placeofPublication, make the change permanent
shake = shake.drop(columns=['doi', 'placeOfPublication'])

In [ ]:
# get the title and the creator of the documents published between 2000 and 2010
filt = (shake['publicationYear']>1999) & (shake['publicationYear']<2011)
shake.loc[filt, ['title', 'creator']]

In [ ]:
# get the creator of the documents shorter than 10 pages or longer than 50 pages
filt = (shake['pageCount']<10)|(shake['pageCount']>50)
shake.loc[filt, 'creator']

In [ ]:
# get the title of the documents whose publisher is not Folger Shakespeare Library
filt = (shake['publisher']=='Folger Shakespeare Library')
shake.loc[~filt, 'title']

In [ ]:
# Make all column names in the Shakespeare dataframe upper case
shake.columns = shake.columns.str.upper()

In [ ]:
# Get all documents whose current title is 'Review Article' and change their title to 'Review'
shake.loc[shake['TITLE']=='Review Article', 'TITLE'] = 'Review'

In [ ]:
# Get all documents whose word count exceeds 5000 and change their word count to the string 'Long article'
shake.loc[shake['WORDCOUNT']>5000, 'WORDCOUNT'] = 'Long article'